File: App/models/Articulo.py
class Articulo:
    def __init__(self, codigo, nombre, presentaciones, categoria, lote, fecha_vencimiento, extras=None):
        self.codigo = codigo
        self.nombre = nombre
        self.presentaciones = presentaciones
        self.categoria = categoria
        self.lote = lote
        self.fecha_vencimiento = fecha_vencimiento
        self.extras = extras if extras else {}  # Atributos adicionales

    def to_mongo(self):
        return {
            "codigo": self.codigo,
            "nombre": self.nombre,
            "presentaciones": self.presentaciones,
            "categoria": self.categoria,
            "lote": self.lote,
            "fecha_vencimiento": self.fecha_vencimiento,
            **self.extras  # Se agregan atributos extra según la categoría
        }

    @staticmethod
    def from_mongo(data):
        return Articulo(
            codigo=data["codigo"],
            nombre=data["nombre"],
            presentaciones=data["presentaciones"],
            categoria=data["categoria"],
            extras={k: v for k, v in data.items() if k not in ["codigo", "nombre", "presentaciones", "categoria"]}
        )

File: App/static/js/articulo.js
document.addEventListener("DOMContentLoaded", function () {
    const form = document.getElementById("form-articulo");

    if (!form) {
        console.error("Error: No se encontró el formulario con id 'form-articulo'. Verifica el HTML.");
        return;
    }

    async function cargarArticulos() {
        try {
            const response = await fetch("/listar_articulos");

            if (!response.ok) {
                console.error("Error al obtener artículos:", response.statusText);
                return;
            }

            const data = await response.json();
            console.log("Artículos obtenidos:", data);

            const tabla = document.getElementById("articulosTable");
            if (!tabla) {
                console.error("Error: No se encontró la tabla con id 'articulosTable'.");
                return;
            }

            tabla.innerHTML = ""; // Limpiar tabla antes de llenarla

            data.articulos.forEach(articulo => {
                const fila = document.createElement("tr");

                fila.innerHTML = `
                    <td>${articulo.codigo}</td>
                    <td>${articulo.nombre}</td>
                    <td>${articulo.presentaciones.join(", ")}</td>
                    <td>${articulo.categoria}</td>
                    <td>${articulo.lote}</td>
                `;

                tabla.appendChild(fila);
            });

        } catch (error) {
            console.error("Error al obtener artículos:", error);
        }
    }

    form.addEventListener("submit", async function (event) {
        event.preventDefault();

        const codigo = document.getElementById("codigo").value.trim();
        const nombre = document.getElementById("nombre").value.trim();
        const presentaciones = document.getElementById("presentaciones").value.trim();
        const categoria = document.getElementById("categoria").value;
        const lote = document.getElementById("lote").value.trim();
        const fecha_vencimiento = document.getElementById("fecha_vencimiento").value;
        const csrfToken = document.querySelector("input[name='csrf_token']")?.value; // Opcional

        if (!codigo || !nombre || !presentaciones || !categoria) {
            alert("Todos los campos son obligatorios.");
            return;
        }

        const articuloData = {
            codigo,
            nombre,
            presentaciones: presentaciones.split(", "),
            categoria,
            lote,
            fecha_vencimiento
        };

        try {
            const response = await fetch("/guardar_articulo", {
                method: "POST",
                headers: {
                    "Content-Type": "application/json",
                    "Accept": "application/json",
                    ...(csrfToken && { "X-CSRFToken": csrfToken }) // Añadir CSRF si está presente
                },
                body: JSON.stringify(articuloData)
            });

            if (!response.ok) {
                const errorText = await response.text();
                console.error("Error del servidor:", errorText);
                alert("Error al guardar el artículo: " + errorText);
                return;
            }

            const data = await response.json();
            console.log("Respuesta del servidor:", data);

            alert("Artículo guardado con éxito.");
            form.reset();
            cargarArticulos();
        } catch (error) {
            console.error("Error al guardar:", error);
            alert("Hubo un problema al conectar con el servidor.");
        }
    });

    cargarArticulos();
});
